packages/core/src/types.ts

```diff
@@ -76,12 +76,29 @@ export interface OperationResult<Data = any> {
 
 /** Input parameters for to an Exchange factory function. */
 export interface ExchangeInput {
-  forward: ExchangeIO;
   client: Client;
+  forward: ExchangeIO;
 }
 
 /** Function responsible for listening for streamed [operations]{@link Operation}. */
 export type Exchange = (input: ExchangeInput) => ExchangeIO;
 
 /** Function responsible for receiving an observable [operation]{@link Operation} and returning a [result]{@link OperationResult}. */
 export type ExchangeIO = (ops$: Source<Operation>) => Source<OperationResult>;
+
+/** Debug event types (interfaced for declaration merging). */
+export interface DebugEventTypes {
```

Comment:
I think strictly typing this defeats the purpose of this, what if a thirth party exchange is used that has a `cache successfully restored` event (hypothetical). Are we expecting them to move to this repository to add that type here and wait for a publish? We could strictly type this in integrations (like devtools) so we avoid having to support all event-types.

Reply:
So, `dispatchEvent` does allow for `string` I believe https://github.com/FormidableLabs/urql/pull/608/files#diff-65274cc049fe6e9b471d39b43d571e44R18

But I think we can address the events themselves in a separate PR, when we implement them. Specifically, `cacheInvalidation` will be different between Graphcache and the document cache, so I'd treat this as tentative events for now that will be replaced when we start adding this to the exchanges (?)